keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+  the problem with node drains and offers a workaround to restart the application without the
+  application owner's consents, but acknowledges that this may be problematic without the knowledge
+  of the application
+- Jack Roper, [How to Delete Pods from a Kubernetes Node with Examples](https://spacelift.io/blog/kubectl-delete-pod), 05 Jul, 2023 - also
+  discusses the problem of blocking PDBs and offers several workarounds. Similar to others also
+  offers a force deletion, but also a less destructive method of scaling up the application.
+  However, this also interferes with application deployment and has to be supported by the
+  application.
+
+### Cluster Autoscaler
+
+Accepts a `drain-priority-config` option, which is similar to Graceful Node Shutdown in that it
+gives each priority a shutdown grace period. Also has a `max-graceful-termination-sec` option for
+pod termination and a `max-pod-eviction-time` option after which the eviction is forfeited.
+
+Each pod is first analyzed to see if it is drainable. Part of the logic is similar to kubectl and
+its drain filters (see [Cluster Autoscaler rules](https://github.com/kubernetes/autoscaler/blob/554366f979b11aeb82df335a793e4d7a1acfadb4/cluster-autoscaler/simulator/drainability/rules/rules.go#L50-L77)):
+- Mirror pods are skipped.
+- Terminating pods are skipped.
+- Pods and ReplicaSets/ReplicationControllers without owning controllers are blocking by default
+  (the check can be modified with the `skip-nodes-with-custom-controller-pods` option).
+- System pods (in the `kube-system` namespace) without a matching PDB are blocking by default
+  (the check can be modified with the `skip-nodes-with-system-pods` option).
+- Pods with `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation are blocking.
+- Pods with local storage are blocking unless they have a
+  `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` annotation
+  (the check can be modified with the `skip-nodes-with-local-storage` option, e.g., this check
+  is skipped on [AKS](https://learn.microsoft.com/en-us/azure/aks/cluster-autoscaler?tabs=azure-cli#cluster-autoscaler-profile-settings)).
+- Pods with PDBs that do not have `disruptionsAllowed` are blocking.
+
+This can be enhanced with other rules and overrides.
+
+It uses this logic to first check if all pods can be removed from a node. If not, it will report
+those nodes. Then it will group all pods by a priority and evict them gradually from the lowest
+to highest priority. This may include DaemonSet pods.
+
+### kubelet
+
+Graceful Node Shutdown is a part of the current solution for node maintenance. Unfortunately, it is
+not possible to rely solely on this feature as a go-to solution for graceful node and workload
+termination.
+
+- The Graceful Node Shutdown feature is not application aware and may prematurely disrupt workloads 
```

Comment:
None of the solutions will be ever application aware, that is one of the reasons we have operators per a specific application, which know how to deal with all the intricacies of managing it.

Reply:
Yes, I am only stating the fact. Not implying that it should be. The point of this KEP is to freely share the intentions of draining a node and each pod so that each actor (controller, application, user, tools, etc.) can react in a proper way.